### DISCUSSION TITLE
Error: There is no X field when I add new fields in my registration form with keycloakify

### DISCUSSION THREAD

**Discussion by kasir-barati** [keycloakify/keycloakify#281] [Q&A] [answered]
Hi dear reader, As you can see [here](https://github.com/kasir-barati/customized-keycloakify-starter) I just happened to add some extra fields to my registration form; location and occupation. I can build and use it. but when I go to the registration page I see an white page with this error in the console:
![image](https://user-images.githubusercontent.com/73785723/228107365-cb2f07e2-f4e6-4870-bbec-c156be8d5d8a.png)
Any help would be super great and useful.

here is the complete version:
```cmd
Some cookies are misusing the recommended “SameSite“ attribute 8
Cookie “KEYCLOAK_IDENTITY” has been rejected because it is already expired. registration
Cookie “KEYCLOAK_IDENTITY_LEGACY” has been rejected because it is already expired. registration
Cookie “KEYCLOAK_SESSION” has been rejected because it is already expired. registration
Cookie “KEYCLOAK_SESSION_LEGACY” has been rejected because it is already expired. registration
Cookie “KEYCLOAK_IDENTITY” has been rejected because it is already expired. registration
Cookie “KEYCLOAK_IDENTITY_LEGACY” has been rejected because it is already expired. registration
Cookie “KEYCLOAK_SESSION” has been rejected because it is already expired. registration
Cookie “KEYCLOAK_SESSION_LEGACY” has been rejected because it is already expired. registration
Error: There is no location field
    printIfExists registration:312
    o Register.tsx:128
    React 8
    x scheduler.production.min.js:13
    A scheduler.production.min.js:14
react-dom.production.min.js:186:401
    React 9
    x scheduler.production.min.js:13
    A scheduler.production.min.js:14
Uncaught Error: There is no location field
    printIfExists registration:312
    o Register.tsx:128
    React 8
    x scheduler.production.min.js:13
    A scheduler.production.min.js:14
registration:312:23
The resource at “http://localhost:8080/resources/pulcx/login/customized-keycloakify-starter/build/fonts/WorkSans/worksans-bold-webfont.woff2” preloaded with link preload was not used within a few seconds. Make sure all attributes of the preload tag are set correctly. registration
The resource at “http://localhost:8080/resources/pulcx/login/customized-keycloakify-starter/build/fonts/WorkSans/worksans-medium-webfont.woff2” preloaded with link preload was not used within a few seconds. Make sure all attributes of the preload tag are set correctly. registration
The resource at “http://localhost:8080/resources/pulcx/login/customized-keycloakify-starter/build/fonts/WorkSans/worksans-regular-webfont.woff2” preloaded with link preload was not used within a few seconds. Make sure all attributes of the preload tag are set correctly. registration
The resource at “http://localhost:8080/resources/pulcx/login/customized-keycloakify-starter/build/fonts/WorkSans/worksans-semibold-webfont.woff2” preloaded with link preload was not used within a few seconds. Make sure all attributes of the preload tag are set correctly. registration
```

**Comment by garronej** [comment, marked as answer]
Oh damn it,  
Sorry I forgot about [this](https://github.com/keycloakify/keycloakify/issues/40#issuecomment-1202102662).  

I'll add this feature tonight.   

In the meantime I've manually added location and occupation. You can update Keycloakiy

**Reply by garronej** [reply]
And by the way `register.ftl` is kind of legacy (keycloak wise), the new register-user-profile.ftl is much more powerfull since it enables https://docs.keycloakify.dev/realtime-input-validation

**Reply by kasir-barati** [reply]
Thanks for the reply. I am truly thankful :heart: but I am kinda a bit lost. Should I use this method whenever I wanna add or remove my custom fields:

```json
"extraFields": [ 
      "user.attributes.linkedinAccount", 
      "user.attributes.hobbies", 
      "user.attributes.nickname" 
]
```

or something else?

**I just used some random fields to clarify what I meant**

**Reply by kasir-barati** [reply]
> And by the way `register.ftl` is kind of legacy (keycloak wise), the new register-user-profile.ftl is much more powerfull since it enables https://docs.keycloakify.dev/realtime-input-validation

Does this mean that I can delete [this part of code](https://github.com/kasir-barati/customized-keycloakify-starter/blob/main/src/login/KcApp.tsx#L65-L76) entirely and keycloak will use `register-user-profile.ftl` automatically?

**Reply by garronej** [reply]
> Should I use this method whenever I wanna add or remove my custom fields:
Yes! But I don't have time to do tu update right now, I'll do it tonight.    

> Does this mean that I can delete [this part of code](https://github.com/kasir-barati/customized-keycloakify-starter/blob/main/src/login/KcApp.tsx#L65-L76) entirely and keycloak will use register-user-profile.ftl automatically?  

Yes, `register.ftl` is to use only if you are using a Keycloak version that is too old and dosen't support User profile features.

**Reply by kasir-barati** [reply]
Hi @garronej 

I just followed what you said but no result, I delete `Register.tsx` and added the new attributes to the `package.json` but no luck. It is not showing new attributes. I appreciate any kind of help.

https://github.com/kasir-barati/customized-keycloakify-starter/commit/d0a017013b8fd45f3010c97b16ad84d7008934b4

**Reply by garronej** [reply]
Hi,  
Adding the attributes in the package.json is for `Register.ftl` and the feature is not yet implemented.  
For register user profil you add the fields dynamically from the keycloak admin wich is great, you don't need to recompile.  
Everything is explained here: https://docs.keycloakify.dev/realtime-input-validation

**Reply by kasir-barati** [reply]
Thanks for answering my question, Actually I removed `register.ftl` from the keycloakify-starter completely. So basically still I need to add new attributes via the GUI and there is no way to do it via Keycloakify or Terraform? I am really in love with code as infrastructure :joy: and I hate to configure things with GUI. I think  you understand me.

**Reply by garronej** [reply]
You're theme isn't the source of truth of how your realm and client is configured.  
On the contrary it's to your theme to adapt to how Keycloak is configured.  

Let's say you only want to alow pepoles with `@mit.edu.com` email addresses, you would add a regexp to the email attribute in your Keycloak admin and this email would be passed over to your theme so we can have realtime frontend validation.  
If you hardcode the regexp in your theme the validation would only be client side but an attacker could very well send an handcrafterd POST request with whatever email and Keycloak would be OK with it.  

User attributes enables you to have only one source of truth: the Keycloak configuration.  

You are of course free to configure your attribute using the JSON editor.  
This can be easily dumped and restored.  

<img width="1243" alt="image" src="https://user-images.githubusercontent.com/6702424/229113226-3b5916a2-f078-4e9e-a6a0-e36e7b29e7fe.png">

**Reply by garronej** [reply]
I updated the documentation about it: https://docs.keycloakify.dev/realtime-input-validation
